yo1k/tic_tac_toe/core/game.py

```diff
@@ -0,0 +1,282 @@
+
+    def set(self, cell: Cell, mark: Mark):
+        self.cells[cell.x][cell.y] = mark
+
+    def get(self, cell: Cell) -> Optional[Mark]:
+        return self.cells[cell.x][cell.y]
+
+    def clean(self):
+        self.cells = Board.empty_cells()
+
+    @staticmethod
+    def size() -> int:
+        return 3
+
+    @staticmethod
+    def empty_cells() -> list[list[Optional[Mark]]]:
+        return [[None for _ in range(Board.size())] for _ in range(Board.size())]
+
+    def __str__(self):
+        return (
+                f"{type(self)}("
+                f"cells={self.cells},")
+
+
+class State:
+    """Full game state which is enough to restore a saved game."""
+
+    def __init__(
+            self,
+            game_rounds: int,
+            players: Sequence[Player],
+            board: Board,
+            phase=Phase.BEGINNING,
+            round_: int = 0,
+            step: int = 0):
+        self.game_rounds = game_rounds
+        self.round = round_
+        assert len(players) == State.players_count(), f"{players}"
+        self.players = players
+        self.step = step
+        self.phase = phase
+        self.board = board
+
+    def turn(self) -> int:
+        """Returns `turn` - the index of the current `Player` in `State.players`.
+
+        `turn` is calculated in such a way as to alternate players order in different
+        `round` while keeping the same `Player` index in the `State.players`.
+        """
+        return (self.step + self.round) % 2
+
+    @staticmethod
+    def players_count() -> int:
+        return 2
+
+    def __str__(self):
+        return (
+                f"{type(self)}("
+                f"game_rounds={self.game_rounds},"
+                f"round={self.round},"
+                f"players={self.players},"
+                f"step={self.step},"
+                f"phase={self.phase},"
+                f"board={self.board})")
+
+
+class Action:
+    def __init__(self, surrender: bool, occupy: Optional[Cell], next_round: bool):
+        assert \
+            (surrender is True and occupy is None and next_round is False) \
+            or (surrender is False and occupy is not None and next_round is False) \
+            or (surrender is False and occupy is None and next_round is True), \
+            f"surrender={surrender}, occupy={occupy}, next_round={next_round}"
+        self.__surrender = surrender
+        self.__occupy = occupy
+        self.__start = next_round
+
+    @staticmethod
+    def new_surrender() -> Action:
+        return Action(True, None, False)
+
+    @staticmethod
+    def new_occupy(cell: Cell) -> Action:
+        return Action(False, cell, False)
+
+    @staticmethod
+    def new_start() -> Action:
+        return Action(False, None, True)
+
+    @property
+    def surrender(self):
+        return self.__surrender
+
+    @property
+    def occupy(self):
+        return self.__occupy
+
+    @property
+    def start(self):
+        return self.__start
```

Comment:
Should the property methods specify the return type?

Reply:
Done.